File: src/main/java/com/fouan/zones/Zone.java
package com.fouan.zones;

import lombok.AllArgsConstructor;
import lombok.Getter;

import java.util.List;

import static lombok.AccessLevel.PRIVATE;

@AllArgsConstructor(access = PRIVATE)
public final class Zone {

    @Getter
    private final Position position;
    private final List<ZoneMarker> markers;

    public static Zone startingZone(Position position) {
        return new Zone(position, List.of(ZoneMarker.STARTING_ZONE));
    }

    public static Zone exitZone(Position position) {
        return new Zone(position, List.of(ZoneMarker.EXIT_ZONE));
    }

    public static Zone zombieSpawnZone(Position position) {
        return new Zone(position, List.of(ZoneMarker.ZOMBIE_SPAWN));
    }

    public static Zone normalZone(Position position) {
        return new Zone(position, List.of(ZoneMarker.NORMAL_ZONE));
    }

    public List<ZoneMarker> getMarkers() {
        return List.copyOf(markers);
    }

    public enum ZoneMarker {
        STARTING_ZONE,
        EXIT_ZONE,
        ZOMBIE_SPAWN,
        NORMAL_ZONE,
    }
}

File: src/main/java/com/fouan/zones/Zone.kt
package com.fouan.zones

class Zone private constructor(
    val position: Position,
    markers: List<ZoneMarker>
) {

    val markers: List<ZoneMarker> = markers
        get() = field.toMutableList()

    enum class ZoneMarker {
        STARTING_ZONE,
        EXIT_ZONE,
        ZOMBIE_SPAWN,
        NORMAL_ZONE
    }

    companion object {
        @JvmField
        val ZONE_POSITION_COMPARATOR: Comparator<Zone> = Comparator.comparing(Zone::position, Position.BOARD_COMPARATOR)

        fun startingZone(position: Position): Zone = Zone(position, listOf(ZoneMarker.STARTING_ZONE))

        fun exitZone(position: Position): Zone = Zone(position, listOf(ZoneMarker.EXIT_ZONE))

        fun zombieSpawnZone(position: Position): Zone = Zone(position, listOf(ZoneMarker.ZOMBIE_SPAWN))

        fun normalZone(position: Position): Zone = Zone(position, listOf(ZoneMarker.NORMAL_ZONE))
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Zone

        return position == other.position
    }

    override fun hashCode(): Int {
        return position.hashCode()
    }

    override fun toString(): String {
        return "Zone(position=$position)"
    }
}
